### PR TITLE
osd: Include dups in copy_after() and copy_up_to()

### PR DESCRIPTION
**Pull request by dzafman**
- [x] References tracker ticket
- [x] ~Updates documentation if necessary~
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by dzafman** [comment]
4 unrelated rados/thrash failures

4 FAIL dzafman-2019-05-02_19:43:04-rados:thrash-wip-zafman-testing-distro-basic-smithi
	3919741 http://tracker.ceph.com/issues/39581
	3919818 ceph_test_rados dumps core
	3919828 Command failed on smithi131 with status 124: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump -f json-pretty'
	3919975 ceph_test_rados dumps core

File: src/osd/PGLog.cc [hunk 1]
```diff
@@ -122,7 +122,7 @@ void PGLog::IndexedLog::trim(
 
   while (!dups.empty()) {
     const auto& e = *dups.begin();
-    if (e.version.version >= earliest_dup_version)
+    if (e.version.version > earliest_dup_version)
```
Comment (jdurgin) [inline src/osd/PGLog.cc]: could you add a unit test for this? unittest_pglog (test/osd/TestPGLog.cc) has some for trim already

File: src/osd/osd_types.cc [hunk 2]
```diff
@@ -4668,21 +4669,58 @@ void pg_log_t::copy_after(const pg_log_t &other, eversion_t v)
     }
     log.push_front(*i);
   }
+  for (auto i = other.log.cbegin(); i != other.log.cend(); ++i) {
```
Comment (jdurgin) [inline src/osd/osd_types.cc]: it looks like the trim portion for both of these is the same, and could extracted into a common helper

**Comment by dzafman** [comment]
@neha-ojha @jdurgin I think I now have a final version of this code.  I'm re-running rados/thrash testing.

**Comment by dzafman** [comment]
@jdurgin I had a racey standalone test, which after being fixed revealed a bug in merge_log_dups().  See 2 new commits.

**Comment by jdurgin** [comment]
now the unittests need updating to compensate for the fix though

**Comment by dzafman** [comment]
@neha-ojha I fixed the commit message because we don't need to trim the dups in _handle_dups().  
a85b7ee add a test case to test copy_after() with dups present.

@jdurgin I put in unittest_pglog correction needed with merge_dup_logs() fix in a combined commit f39ca3e

rados and addition rados/thrash tests in progress.

**Comment by jdurgin** [comment]
thanks, updates look good

**Comment by dzafman** [comment]
2 runs passed
dzafman-2019-05-09_20:06:24-rados-wip-zafman-testing-distro-basic-smithi
3943731 osd.1 cored but nothing in log about it
3943768 valgrind
3943818 mgr crash rbd related
3943864 valgrind
3943886 mgr test failure
3943959 valgrind
3943830 DEAD infrastructure

dzafman-2019-05-10_07:50:18-rados:thrash-wip-zafman-testing-distro-basic-smithi
3946267 DEAD infrastructure
3946199 infrastructure
3 still running

**Pull request merged**